### PR TITLE
Adding Jensen habit-predicting microphysics to WRF

### PR DESCRIPTION
**Pull request by graupely**
TYPE: new feature

KEYWORDS: New microphysics option, Jensen microphysics, shape, aspect ratio

SOURCE: Anders Jensen (NCAR-RAL)

DESCRIPTION OF CHANGES: 
Adding Jensen microphysics to WRF. This pull request includes the updated framework (changes to Registry, etc.) as well as coupling to the RRTMG radiation. 

The microphysics predicts two ice particle dimensions, and therefore two predicted and advected variables are needed for ice shape. These are the volume mixing ratio (QVOLI) and the aspect ratio times the volume (QAOLI) mixing ratio in the Registry. Additionally mass (QICE) and number (QNICE) mixing ratios are predicted.  Four prognostic variables are need for each ice species, and there are three ice species (planar habits, column habits, aggregates) which makes twelve prognostic ice variables. I used the same framework/notation as used for P3 (e.g., QICE, QICE2, QICE3).  I also have similar diagnostics as used in P3 with added ice particle aspect ratio (phi) and an ice type diagnostic that will link particle properties back to traditional ice types such as cloud ice, snow, graupel, and hail. 

The lookup tables used are large; I had to split the ice-rain collection lookup table to upload it to github. Therefore, there are four lookup tables. Two handle ice-rain collection, one does ice-cloud water collection, and one calculates gamma functions for various particle shapes.           

LIST OF MODIFIED FILES:
M Makefile
M Registry/Registry.EM_COMMON
M dyn_em/module_first_rk_step_part1.F
M dyn_em/solve_em.F
M phys/Makefile
M phys/module_microphysics_driver.F
A phys/module_mp_jensen_ishmael.F
M phys/module_physics_init.F
M phys/module_radiation_driver.F
A run/ishmael-gamma-tab.dat
A run/ishmael-qi-qc.dat
A run/ishmael-qi-qr_1.dat
A run/ishmael-qi-qr_2.dat

TESTS CONDUCTED: 
1. A small GNU-only WTF has been done successfully. 
2. Restart gives bit-wise identical results.
3. We recently tested a 3D squall-line (setup using em_quarter_ss), and a winter-time orographic precipitation case which are described:
https://journals.ametsoc.org/doi/abs/10.1175/JAS-D-17-0400.1
https://journals.ametsoc.org/doi/10.1175/MWR-D-17-0215.1
4. Before this pull request I also tested ifort and gnu compilers. 
5. I have tested the scheme using 4 nests (12km, 4km, 1.3km, 148m) for a lake effect snow case (publication in the works).

### CODE REVIEW

**Comment by davegill** [comment]
@graupely 
Would you modify the "base" branch to be _develop_ and not _master_. To do this, you probably just need to click on the edit button (next to the commit message title) and change _wrf-model-master_ to _wrf-model:develop_.

Watch out that the number of files still stays at what you expect.

**Comment by davegill** [comment]
@graupely 
Also, on the "SOURCE" line, modify this to a conventional 
SOURCE: Anders Jensen (affiliation)

**Comment by davegill** [comment]
@graupely 
Not in the commit message, but here would suffice - how big are each of the lookup data files?

**Comment by graupely** [comment]
@davegill The ice-rain collection lookup table is the largest at ~140 MB. Ice-cloud water and the gamma tables are smaller ~30-40 MB.

**Comment by dudhia** [comment]
Yes, we should find a better number.

**Comment by davegill** [comment]
@graupely 
 - [x] Used the package option
 - [x] Variables are declared optional
 - [x] Variables are checked for present
 - [x] Variables are checked for f_xxx before use
 - [x] Local computations are tile sized

File: Registry/Registry.EM_COMMON [hunk 1]
```diff
@@ -2649,8 +2692,10 @@ package   p3_1category    mp_physics==50               -             moist:qv,qc
 package   p3_1category_nc mp_physics==51               -             moist:qv,qc,qr,qi;scalar:qnc,qni,qnr,qir,qib;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,refl_10cm,th_old,qv_old
 package   p3_2category    mp_physics==52               -             moist:qv,qc,qr,qi,qi2;scalar:qnc,qni,qnr,qir,qib,qni2,qir2,qib2;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,vmi3d_2,rhopo3d_2,di3d_2,refl_10cm,th_old,qv_old
 package   morr_tm_aero    mp_physics==40               -             moist:qv,qc,qr,qi,qs,qg;scalar:qnc,qni,qns,qnr,qng;state:rqrcuten,rqscuten,rqicuten,EFCG,EFIG,EFSG,WACT,CCN1_GS,CCN2_GS,CCN3_GS,CCN4_GS,CCN5_GS,CCN6_GS,CCN7_GS,re_cloud,re_ice,re_snow,mskf_refl_10cm
+package   jensen_ishmael  mp_physics==86               -             moist:qv,qc,qr,qi,qi2,qi3;scalar:qnr,qni,qvoli,qaoli,qni2,qvoli2,qaoli2,qni3,qvoli3,qaoli3;state:re_cloud,re_ice,vmi3d,rhopo3d,di3d,phii3d,itype,vmi3d_2,rhopo3d_2,di3d_2,phii3d_2,itype_2,vmi3d_3,rhopo3d_3,di3d_3,phii3d_3,itype_3,refl_10cm
```
[the remark this answers is not in the document]
Comment (graupely) [reply Registry/Registry.EM_COMMON]: My bad. I was trying to go high and stay out of peoples' way. How about 33?  It looks like no one has the 30s, though Greg will likely use 38.

**Comment by dudhia** [comment]
The 30's are currently bin schemes which are in a different registry file.
Looks like room in the 20's (e.g. 25).

**Comment by graupely** [comment]
I always forget about the bins schemes... 
Maybe I should jump into the 50s with Hugh.  He has 50-52. Could I take 54 or 55?

**Comment by weiwangncar** [comment]
@davegill Should consult with Carlos to see if they would like to replace the old Goddard scheme (option 7). We can then move old Goddard to 97, one of the new Goddard to 7, and the other to 27 perhaps.

**Comment by dudhia** [comment]
We will use 55 for jensen-ishmael as the Goddard schemes were moved
elsewhere.

**Comment by smileMchen** [comment]
@graupely @davegill 
WRF-NMM cannot compile with the newly added Jensen microphysics.

**Comment by graupely** [comment]
@smileMchen I have fixed the issue and wrf-nmm will now compile.

**Comment by smileMchen** [comment]
@graupely 
Thank you for the fix. Another issue is, would you please add the following lines to Makefile after line 560 (in the top WRF directory):
               ln -sf ../../run/ishmael-gamma-tab.dat . ;              \    
               ln -sf ../../run/ishmael-qi-qc.dat . ;                         \    
               ln -sf ../../run/ishmael-qi-qr_1.dat . ;                        \    
               ln -sf ../../run/ishmael-qi-qr_2.dat . ;                        \   

This is to make sure these data files are available when running with Jensen microphysics.

**Comment by graupely** [comment]
> @graupely
> Thank you for the fix. Another issue is, would you please add the following lines to Makefile after line 560 (in the top WRF directory):
> ln -sf ../../run/ishmael-gamma-tab.dat . ; \
> ln -sf ../../run/ishmael-qi-qc.dat . ; \
> ln -sf ../../run/ishmael-qi-qr_1.dat . ; \
> ln -sf ../../run/ishmael-qi-qr_2.dat . ; \
> 
> This is to make sure these data files are available when running with Jensen microphysics.

@smileMchen: issue resolved.

File: phys/module_mp_jensen_ishmael.F [hunk 2]
```diff
@@ -0,0 +1,4701 @@
...
+  !.. Read in ice-cloud water collection lookup table
+    if(.not.allocated(itab)) then 
+       allocate(itab(51,51,51,11,2))
+
+       open(20, FILE='ishmael-qi-qc.dat')
+       do i1 = 1,51
+          do i2 = 1,51
+             do i3 = 1,51
+                do i4 = 1,11
+                   read(20, *) itab(i1,i2,i3,i4,1), &
```
Comment (mgduda) [inline phys/module_mp_jensen_ishmael.F]: The read statements here imply that the look-up tables are stored as formatted text, right? On the one hand, this avoids any endian issues when reading on various architectures, but it does mean that the files will likely be several times larger than they would be if the data were stored in binary.

**Comment by smileMchen** [comment]
@dudhia 
You are right. I didn't change the format of these look-up tables.

File: phys/module_mp_jensen_ishmael.F (hunk 2 above)
Comment (graupely) [reply phys/module_mp_jensen_ishmael.F]: @mgduda I can convert to binary. Is there an easy way to ensure proper endian?

Comment (mgduda) [reply phys/module_mp_jensen_ishmael.F]: I think WRF expects all unformatted files to be in big-endian format. In this case, when writing the unformatted look-up tables, you can compile your code with a byteswap flag that is compiler-dependent:
* ifort: `-convert big_endian`
* gfortran: `-fconvert=big-endian`
* pgf90: `-byteswapio`

**Comment by smileMchen** [comment]
@graupely 
The continuous run and restart run cannot yield identical results when using the Jensen microphysics. Would you please take a look? 
(By the way, would you please also committ the revised code module_mp_jensen_ishmael.F that reads the binary look-up tables? Thanks.)

**Comment by graupely** [comment]
> @graupely
> The continuous run and restart run cannot yield identical results when using the Jensen microphysics. Would you please take a look?
> (By the way, would you please also committ the revised code module_mp_jensen_ishmael.F that reads the binary look-up tables? Thanks.)

I have committed the changes to module_mp_jensen_ishamael.F and the main Makefile. I'll work on debugging the restart issue.  Thanks.

**Comment by davegill** [comment]
@weiwangncar @dudhia @smileMchen @kkeene44 
Michael raises a good point. We should probably manually commit this PR from the command line. Likely one of us does the commit, and we should remember to NOT do the squash and merge option.

**Comment by mgduda** [comment]
@davegill Actually, I think squash-and-merge may be what we want; or we could do an interactive rebase to manually squash the addition of the ASCII tables and the removal of the ASCII tables, for a net change of zero. Not squashing any commits would probably lead to a non-trivial increase in repository size.

**Comment by graupely** [comment]
@smileMchen 
I have uploaded a new version of the microphysics that will hopefully give same answers for continuous versus restarted runs. Let me know. Thanks, Anders

**Comment by smileMchen** [comment]
@graupely 

Thank you very much!

**Comment by smileMchen** [comment]
@graupely 

I am trying to run regression using your version of code. However, every time I try to compile the code, I got the message below:
```
*****************************************************************************
************************** W A R N I N G ************************************
NETCDF4 IO features are requested, but this installation of NetCDF           
  /glade/u/apps/ch/opt/netcdf/4.6.1/gnu/6.3.0
DOES NOT support these IO features.                                          

Please make sure NETCDF version is 4.1.3 or later and was built with         
--enable-netcdf4                                                             

OR set NETCDF_classic variable                                               
   bash/ksh : export NETCDF_classic=1
      echo  csh : --enable-netcdf4 

Then re-run this configure script                                            

!!! configure.wrf has been REMOVED !!!
```
Is the above option "--enable-netcdf4"  or  "--enable-netcdf4 " still necessary?

**Comment by smileMchen** [comment]
@graupely 
Before you commit the PR to change the format of look-up table, the code can be compiled successfully.
I just looked through the codes you committed, but didn't find any changes that may lead to the configure changes?

**Comment by graupely** [comment]
> @graupely
> Before you commit the PR to change the format of look-up table, the code can be compiled successfully.
> I just looked through the codes you committed, but didn't find any changes that may lead to the configure changes?

@smileMchen 
The version of WRF that I have on Cheyenne with my most recent updates configures correctly. When I clone the most recent version from github, I get the same configure errors that you get. I'm a bit confused and I am wondering if there was a change that went in after my microphysics that broke the code?

**Comment by davegill** [comment]
@smileMchen @graupely 
Folks,
Short answer: In the window that you are going to do your regtest, set an environment variable:
```
setenv NETCDF_classic 1
```
or
```
export NETCDF_classic=1
```

**Comment by smileMchen** [comment]
@graupely @davegill 
Just want you to know that the regression test is done successfully. The issue we have previously is caused by an extra empty line in Makefile (thanks to Dave). After deleting that line, everything works fine.  
We are now working on the issue of single versus double precision runs. The look-up tables, I suppose, are single precision binary data. I will keep you updated.

**Comment by graupely** [comment]
> @graupely @davegill
> Just want you to know that the regression test is done successfully. The issue we have previously is caused by an extra empty line in Makefile (thanks to Dave). After deleting that line, everything works fine.
> We are now working on the issue of single versus double precision runs. The look-up tables, I suppose, are single precision binary data. I will keep you updated.

@smileMchen, good news. Yes, the lookup tables are single precision.

**Comment by davegill** [comment]
@graupely @smileMchen @dudhia @weiwangncar 
Folks,
Outstanding issues to resolve:
1. Numbering: looks like we change 86 -> 55.
2. What is the status of restart. A "pass" is not necessary for commit, but we need to know
3. What is the status of using a double precision build with the single precision data. A "pass" is not necessary, but we need to know.
4. Resolve conflicts in physics init.

**Comment by graupely** [comment]
> @graupely @smileMchen @dudhia @weiwangncar
> Folks,
> Outstanding issues to resolve:
> 
> 1. Numbering: looks like we change 86 -> 55.
> 2. What is the status of restart. A "pass" is not necessary for commit, but we need to know
> 3. What is the status of using a double precision build with the single precision data. A "pass" is not necessary, but we need to know.
> 4. Resolve conflicts in physics init.

@davegill 
1. Thanks! 55 is a good number.
2. I tested this myself and it passed, though I am unsure of the official status.
3. I have never tested this. We may need to chat if this fails. 
4. Resolved.

**Comment by smileMchen** [comment]
@graupely
Dave has approved this PR. Please merge it into develop branch. I will take care of the look-up table issue after I am back from the tutorial trip.

**Comment by davegill** [comment]
@graupely @smileMchen 
Anders,
I'll merge this code in.

**Comment by graupely** [comment]
Thanks @davegill. I'll can run a double precision test @smileMchen.

**Pull request merged**